File: src/main/java/ai/aimachineserver/domain/games/tictactoe/GameTicTacToe.java
package ai.aimachineserver.domain.games.tictactoe;

import ai.aimachineserver.domain.games.AbstractGame;
import ai.aimachineserver.domain.games.KeyValue;
import org.json.JSONObject;
import org.springframework.web.socket.WebSocketSession;

public class GameTicTacToe extends AbstractGame {
    private static final Player playerStub = new PlayerHuman("", Symbol.SYMBOL_X);
    private final Board board;
    private final Judge judge;

    private Player player1 = playerStub;
    private Player player2 = playerStub;
    private Player currentPlayer = player1;
    private TurnResult turnResult = TurnResult.GAME_ONGOING;
    private int turnNumber = 0;

    public GameTicTacToe(Board board, Judge judge, String gameName) {
        super(gameName);
        this.board = board;
        this.judge = judge;
    }

    @Override
    public void onPlayerJoinedGame(WebSocketSession session) {
        playerSessions.add(session);
        if (player1 == playerStub) {
            player1 = new PlayerHuman(session.getId(), Symbol.SYMBOL_O);
            currentPlayer = player1;
        } else {
            player2 = new PlayerHuman(session.getId(), Symbol.SYMBOL_X);
            broadcastMessage(new KeyValue("eventType", "current_player"), new KeyValue("eventMessage", currentPlayer.name));
        }
        int playersCount = playerSessions.size();
        broadcastMessage(new KeyValue("eventType", "players_count"), new KeyValue("eventMessage", "$playersCount"));
        String message = playersCount == 1 ? "Waiting for opponent" : "Game has started";
        broadcastMessage(new KeyValue("eventType", "server_message"), new KeyValue("eventMessage", message));
    }

    @Override
    public void onFieldClicked(int rowIndex, int colIndex) {
        if (board.isFieldAvailable(rowIndex, colIndex)) {
            System.out.printf("Clicked [row, col]: [%d, %d]", rowIndex, colIndex);
            JSONObject json = new JSONObject();
            json.put("rowIndex", rowIndex);
            json.put("colIndex", colIndex);
            json.put("fieldToken", currentPlayer.symbol.token);
            String data = json.toString();
            broadcastMessage(new KeyValue("eventType", "new_move_to_mark"), new KeyValue("eventMessage", data));
            if (turnResult == TurnResult.GAME_ONGOING) {
                turnNumber++;
                currentPlayer.makeMove(board, rowIndex, colIndex);
                turnResult = judge.announceTurnResult(turnNumber);
                if (turnResult == TurnResult.GAME_ONGOING) {
                    currentPlayer = assignPlayer();
                    broadcastMessage(new KeyValue("eventType", "current_player"), new KeyValue("eventMessage", currentPlayer.name));
                } else {
                    String resultMessage = getResultMessage();
                    broadcastMessage(new KeyValue("eventType", "game_ended"), new KeyValue("eventMessage", resultMessage));
                    broadcastMessage(new KeyValue("eventType", "current_player"), new KeyValue("eventMessage", "none"));
                }
            }
        }
    }

    private Player assignPlayer() {
        if (currentPlayer == player1) {
            return player2;
        } else {
            return player1;
        }
    }

    private String getResultMessage() {
        if (turnResult == TurnResult.TIE) {
            return "Tie";
        } else {
            return "${currentPlayer.symbol.identifier} won";
        }
    }

    @Override
    public void onDisconnect(WebSocketSession session) {
        playerSessions.remove(session);
        broadcastMessage(new KeyValue("eventType", "server_message"), new KeyValue("eventMessage", "Player disconnected"));
        broadcastMessage(new KeyValue("eventType", "server_message"), new KeyValue("eventMessage", playerSessions.size() + " players in game"));
    }
}

File: src/main/kotlin/ai/aimachineserver/domain/games/tictactoe/GameTicTacToe.kt
package ai.aimachineserver.domain.games.tictactoe

import ai.aimachineserver.domain.games.AbstractGame
import org.json.JSONObject
import org.springframework.web.socket.WebSocketSession

class GameTicTacToe(
    private val board: Board = Board(),
    private val judge: Judge = Judge(board),
    override val gameName: String
) : AbstractGame() {
    private companion object {
        val playerStub = PlayerHuman("", Symbol.SYMBOL_X)
    }

    private var player1 = playerStub
    private var player2 = playerStub
    private var currentPlayer = player1
    private var turnResult = TurnResult.GAME_ONGOING
    private var turnNumber = 0
    override val playerSessions = mutableListOf<WebSocketSession>()

    override fun getAllPlayerSessions() = playerSessions

    override fun onPlayerJoinedGame(session: WebSocketSession) {
        playerSessions.add(session)
        if (player1 == playerStub) {
            player1 = PlayerHuman(session.id, Symbol.SYMBOL_O)
            currentPlayer = player1
        } else {
            player2 = PlayerHuman(session.id, Symbol.SYMBOL_X)
            broadcastMessage("eventType" to "movement_allowed", "eventMessage" to currentPlayer.name)
        }
        val playersCount = playerSessions.count()
        broadcastMessage(
            "eventType" to "server_message",
            "eventMessage" to "$playersCount players in game"
        )
        val message = if (playersCount == 1) "Waiting for opponent" else "Game has started"
        broadcastMessage("eventType" to "server_message", "eventMessage" to message)
    }

    override fun onFieldClicked(rowIndex: Int, colIndex: Int) {
        if (board.isFieldAvailable(rowIndex, colIndex)) {
            println("Clicked [row, col]: [$rowIndex, $colIndex]")
            val data = JSONObject()
                .put("rowIndex", rowIndex)
                .put("colIndex", colIndex)
                .put("fieldToken", currentPlayer.symbol.token)
                .toString()
            broadcastMessage("eventType" to "field_to_be_marked", "eventMessage" to data)
            if (turnResult == TurnResult.GAME_ONGOING) {
                turnNumber++
                currentPlayer.makeMove(board, rowIndex, colIndex)
                turnResult = judge.announceTurnResult(turnNumber)
                if (turnResult == TurnResult.GAME_ONGOING) {
                    changePlayer()
                } else {
                    val resultMessage = getResultMessage()
                    broadcastMessage(
                        "eventType" to "server_message",
                        "eventMessage" to "Game has ended: $resultMessage"
                    )
                    broadcastMessage("eventType" to "movement_allowed", "eventMessage" to "none")
                }
            }
        }
    }

    private fun changePlayer() {
        currentPlayer = if (currentPlayer == player1) {
            player2
        } else {
            player1
        }
        broadcastMessage("eventType" to "movement_allowed", "eventMessage" to currentPlayer.name)
    }

    private fun getResultMessage() = if (turnResult == TurnResult.TIE) {
        "Tie"
    } else {
        "${currentPlayer.symbol.identifier} won"
    }

    override fun onDisconnect(session: WebSocketSession) {
        playerSessions.remove(session)
        broadcastMessage("eventType" to "server_message", "eventMessage" to "Player disconnected")
        broadcastMessage(
            "eventType" to "server_message",
            "eventMessage" to "${playerSessions.count()} players in game"
        )
    }
}
